src/jaegertracing/SpanContext.h

```diff
@@ -45,6 +52,7 @@ class SpanContext : public opentracing::SpanContext {
         , _spanID(0)
         , _parentID(0)
         , _flags(0)
+        , _tracerType(TracerType::TRACER_TYPE_UNKNOWN)
```

Comment:
Why is this needed? The usual approach is to configure tracer with one type of codecs, which apply to both inbound and outbound requests. There’s no need to remember what type of encoding was used in the inbound.

Reply:
But how and where do we need to configure a particular type of codec and based on what?
Here in this PR, I've based it on the incoming request headers. As it would be either B3 / Uber, one of the tracer type would be set. If the trace is initiated by that node itself, it would be set with options in startSpanWithOptions API. 
If we dont remember type of tracing then using injectSpan, we need to have some flag to indicate what type of tracing should be done so that appropriate headers can be set.

Reply:
See https://github.com/jaegertracing/jaeger-client-java/pull/694/files, specifically `CodecConfiguration.withPropagation`

Reply:
@yurishkuro thanks for the reference, I still fail to understand on how this implementation can be used dynamically. As per readme:
```
b3Codec = new B3TextMapCodec();
tracer = new JaegerTracer.Builder(serviceName)
  .registerInjector(Format.Builtin.HTTP_HEADERS, b3Codec)
  .registerExtractor(Format.Builtin.HTTP_HEADERS, b3Codec)
  ...
```
I also checked in stagemonitor.git, looks like application is/have to build a object at start by specifying the tracing type e.g. B3 / Uber. 

If this is so then, application is aware what kind of headers it is expecting and there wouldn't be a need to remember the tracing type. 

If I'm missing something, please help point it out.

Reply:
The code snippet you showed is exactly how you do it. The tracer is created once at application start-up, and at that time you can replace the codecs. After that no more "dynamic" configuration is needed, nor does the span need to remember which codec is used.

Reply:
oh ok, which means that application has to be aware of what type tracing type has to supported explicitly which can be either Uber or B3 mutually exclusively.

Reply:
Yes, but the type of context used on the wire is usually the same for the whole site, not varied per application or per request.

It is theoretically possible to write a composite codec combined from both Uber and B3 headers, if it's required for backwards compatibility. But a single format for the whole system is highly recommended.

Reply:
Sure thanks, I had made the changes to handle kind of dynamic selection expecting either Uber or B3 headers at runtime at the node. An API getTracerType checks for incoming headers and uses particular tracer type - Uber/B3.

Reply:
Sorry, but I don't think we want to support this "dynamic" behavior. None of the other Jaeger clients do it this way, and I have not seen anyone ever asking for this. The codec is configured when creating a tracer, it does not change after that. A composite codec can be implemented ([example in java](https://github.com/jaegertracing/jaeger-client-java/blob/dfb9d8b9dc849f2e6f880b44b26fb77dfc894a02/jaeger-core/src/main/java/io/jaegertracing/internal/propagation/CompositeCodec.java)), which handles more than one encoding, but it is also doing them _always_, not dynamically. As a result, there's never a need to store "type" in the span context.

I don't know why you're trying to reinvent the wheel. I pointed you to several examples how codecs are done in Jaeger clients in other languages, it's an established pattern.

Reply:
Not being certain of the ingress nodes and its supported types(e.g. Istio currently supports only B3) we thought being flexible to handle any type based on incoming headers would be good. 
Yes, as per the case you mentioned of static provisioning this PR would need more enhancement. Thanks for all the help and references. I shall probably create a new one and close this.